components/tidb_query/src/batch/runner.rs

```diff
@@ -440,16 +424,126 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
             }
 
             // Grow batch size
-            if batch_size < BATCH_MAX_SIZE {
-                batch_size *= BATCH_GROW_FACTOR;
-                if batch_size > BATCH_MAX_SIZE {
-                    batch_size = BATCH_MAX_SIZE
-                }
-            }
+            grow_batch_size(&mut batch_size);
         }
     }
 
     pub fn collect_storage_stats(&mut self, dest: &mut SS) {
         self.out_most_executor.collect_storage_stats(dest);
     }
+
+    pub fn handle_streaming_request(
+        &mut self,
+    ) -> Result<(Option<(StreamResponse, IntervalRange)>, bool)> {
+        let mut warnings = self.config.new_eval_warnings();
+
+        // if the data read finished, is_drained means "finished"
+        let (mut record_len, mut is_drained) = (0, false);
+        let mut chunk = Chunk::default();
+
+        // record count less than batch size and is not drained
+        while record_len < self.stream_min_rows_each_iter && !is_drained {
+            let (drained, cnt) =
+                self.internal_handle_request(self.stream_batch_size, &mut chunk, &mut warnings)?;
+            record_len += cnt;
+            is_drained = drained;
+
+            // Grow batch size
+            grow_batch_size(&mut self.stream_batch_size)
+        }
+
+        if !is_drained || record_len > 0 {
```

Comment:
what if `is_drained && record_len > 0`?

Reply:
> is_drained && record_len > 0

what if `is_drained && record_len == 0`,  it will return `Ok((None, true))`. 

If `is_drained && record_len > 0`, it returns the ` ((data, range), true)` because it's drained.